[frontend] Improve function def parsing with stacked decorators.

zhxchen17 opened the pull request:
Under certain conditions where we apply another decorator on top of triton.jit, there is a small chance that in the decorator body there's a "def" substring in it. This is problematic when JITFunction parses function def using str.find('def'). This causes an issue compiling some pytorch model when we happened to decorate a jit function with a hash string argument which has "def" in it.

This diff tries a bit harder to locate the real "def" position using a more complex regex which should further reduce the chance to hit this error.

jlebar commented on python/test/unit/language/test_decorator.py:
```diff
@@ -0,0 +1,24 @@
+        filename=__file__,
+        inductor_meta={'backend_hash': '8cde1805b523918efde951fadd02ae7fb36906d8f746f2adef0aeffabe53b3f8'},
+        min_elem_per_thread=0
+    )
```
can we minimize this testcase a little and add a brief comment explaining what's being tested?  In particular this is nonobvious because the "def" in question seems to be inside the random hash.

jlebar commented on python/triton/runtime/jit.py:
```diff
@@ -478,7 +479,7 @@ def __init__(self, fn, version=None, do_not_specialize=None, debug=None, noinlin
         # function source code (without decorators)
         self.src = textwrap.dedent(inspect.getsource(fn))
-        self.src = self.src[self.src.find("def"):]
+        self.src = self.src[re.search(r"def (\w+)\s*\((.*?)\):", self.src).start():]
```
Can we remove the parens around `(.*?)`?

Also if we're not in multiline mode I think this does not work, because `.` only matches on the current line?

zhxchen17 replied:
yeah you're right, sorry I forgot the multiline case.

jlebar reviewed (commented):
Well this is embarrassing.  Thank you for the patch.

zhxchen17 commented:
Addressed comments to minimize the test. Tweak the regex a little bit so that we always match against the beginning of a line, which is simpler to reason about.

jlebar commented on python/triton/runtime/jit.py:
```diff
@@ -478,7 +479,7 @@ def __init__(self, fn, version=None, do_not_specialize=None, debug=None, noinlin
         # function source code (without decorators)
         self.src = textwrap.dedent(inspect.getsource(fn))
-        self.src = self.src[self.src.find("def"):]
+        self.src = self.src[re.search(r"^def\s+\w+\s*\(", self.src, re.MULTILINE).start():]
```
Oh I guess the `\w` is okay because we only match the name up until the open paren.

But why is `^` correct?

zhxchen17 replied:
Yeah I'm not matching against arg list anymore, because we probably just need to find the position of `def`. \w+ part correspond to the function name before the first ( in the parsing. `def` should always occur at the beginning of the line, otherwise it will be a indent error in Python. Also `def` is a keyword so no other syntax rule can use it.

zhxchen17 replied:
I was tempting to make it just re.search("^def ", ...) which I couldn't come up with a counter example, but to make a bit safer I add a bunch of further matching again the `foo_name(` part.

zhxchen17 replied:
Oh one thing I forgot to bring up: we always call textwrap.dedent() before this line, and since we always start a `def` with a new line, I feel `^` works. Happy to be corrected.

zhxchen17 commented:
Re-push this to fix yapf lint error.
